stm32 entropy fix

niedzwiecki-dawid opened the pull request:
Enable the RNG module for the get_entropy_isr function if needed.

Fixes #66312

niedzwiecki-dawid committed:
entropy: stm32: enable rng module if needed

Move enabling/disabling the RNG module to acquire/release functions.
It causes enabling the RNG module for the get_entropy_isr function.

It fixes hanging in the get_entropy_isr function.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -595,13 +597,18 @@ static int entropy_stm32_rng_get_entropy_isr(const struct device *dev,
+			entropy_stm32_resume(dev);
+		}
 		rng_already_acquired = z_stm32_hsem_is_owned(CFG_HW_RNG_SEMID);
 		acquire_rng();
```
It should work. Looks cleaner now :-)

erwango commented:
Found some remaining failing tests with current solution:
```
   tests/crypto/rand32/crypto.rand32, nucleo_f091rc (Regression)
   tests/crypto/rand32/crypto.rand32, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_ctr_drbg, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_hw_xoshiro, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_sw_systimer, nucleo_f091rc (Regression)
   tests/crypto/rand32/crypto.rand32.random_sw_systimer, nucleo_h743zi
   tests/crypto/tinycrypt/crypto.tinycrypt, nucleo_f091rc (Regression)
   tests/crypto/tinycrypt/crypto.tinycrypt, nucleo_h743zi
   tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random, nucleo_f091rc (Regression)
   tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random, nucleo_h743zi
   tests/kernel/mem_protect/stackprot/kernel.memory_protection.stackprot, nucleo_f091rc (Regression)
   tests/kernel/mem_protect/sys_sem/kernel.memory_protection.sys_sem.nouser, nucleo_f091rc (Regression)
   tests/net/lib/dns_resolve/net.dns.resolve, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve.no_ipv6, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve.preempt, nucleo_h743zi
```
Tested with `twister --ninja --runtime-artifact-cleanup -e logging -e cmsis_dsp -T tests/crypto -T tests/kernel/mem_protect -T tests/net/lib/dns_resolve/`
Boards are the ones from our test farm (you can see the list here: https://github.com/zephyrproject-rtos/test_results/pull/1788/files)

niedzwiecki-dawid commented:
> tests/kernel/mem_protect/stackprot/kernel.memory_protection.stackprot, nucleo_f091rc (Regression)
>    tests/kernel/mem_protect/sys_sem/kernel.memory_protection.sys_sem.nouser, nucleo_f091rc (Regression)

Can I somehow run tests on these boards or at least see tests logs?

erwango commented:
> Can I somehow run tests on these boards or at least see tests logs?

I'm running a new test session. Note that this target is also impacted by https://github.com/zephyrproject-rtos/zephyr/issues/66388 and additionally was maybe in a bad state. I'll send new results when received

niedzwiecki-dawid commented:
Ok, thanks. I've ordered nucleo_h743zi anyway and should be able to run tests locally in 1/2 days.

erwango commented:
New session results:
```
   tests/crypto/rand32/crypto.rand32, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_ctr_drbg, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_hw_xoshiro, nucleo_h743zi
   tests/crypto/rand32/crypto.rand32.random_sw_systimer, nucleo_h743zi
   tests/crypto/tinycrypt/crypto.tinycrypt, nucleo_h743zi
   tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve.no_ipv6, nucleo_h743zi
   tests/net/lib/dns_resolve/net.dns.resolve.preempt, nucleo_h743zi
 ```

erwango commented:
Back trace when running `tests/crypto/rand32` on a H7 board:
```
generate_from_isr (len=1, buf=0x2400098c <ztest_thread_stack+972> "") at /local/mcu/zephyrproject/zephyr/drivers/entropy/entropy_stm32.c:364
364				__SEV();
(gdb) bt
#0  generate_from_isr (len=1, buf=0x2400098c <ztest_thread_stack+972> "")
    at /local/mcu/zephyrproject/zephyr/drivers/entropy/entropy_stm32.c:364
#1  entropy_stm32_rng_get_entropy_isr (dev=<optimized out>, 
    buf=0x2400098c <ztest_thread_stack+972> "", len=<optimized out>, flags=<optimized out>)
    at /local/mcu/zephyrproject/zephyr/drivers/entropy/entropy_stm32.c:608
#2  0x0800385a in entropy_get_entropy_isr (dev=<optimized out>, flags=1, length=1, 
    buffer=0x2400098c <ztest_thread_stack+972> "")
    at /local/mcu/zephyrproject/zephyr/include/zephyr/drivers/entropy.h:116
#3  z_early_rand_get (buf=buf@entry=0x2400098c <ztest_thread_stack+972> "", length=length@entry=1)
    at /local/mcu/zephyrproject/zephyr/kernel/init.c:517
#4  0x0800060c in rand32_common_test_rand32 ()
```

niedzwiecki-dawid commented:
> Member

Ok, thanks. So the problem occurs only on H7, right?

erwango commented:
> Ok, thanks. So the problem occurs only on H7, right?

Indeed

niedzwiecki-dawid commented:
I have to run the tests on device. I'm not able to find the issue without actual debug.

niedzwiecki-dawid commented:
@erwango Would you be able to retest with the current version?

erwango commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -143,6 +145,7 @@ static int entropy_stm32_resume(const struct device *dev)
 	res = clock_control_on(dev_data->clock,
 			(clock_control_subsys_t)&dev_cfg->pclken[0]);
 	LL_RNG_Enable(rng);
+	LL_RNG_EnableIT(rng);
```
Looks it solves the on going issues and hence a welcome addition, but is there a proper explanation ?

niedzwiecki-dawid replied:
I did some testing and it looks like the IE bit in CR register is sometimes cleared when the device clock is disabled. It looks like it happens only on H7 chip. Unfortunately, I haven't fine any confirmation of such behavior in RM.

erwango commented:
> Would you be able to retest with the current version?

All tests issues are now resolved

erwango reviewed (changes requested):
One question and it would be nice to update commit message to have more information on the changes that were performed.
Otherwise LGTM

niedzwiecki-dawid committed:
entropy: stm32: enable interrupt when RNG is resumed

Enable the RNG module interrupt every time it is resumed.

It is done to make sure the interrupt is always enabled. The CR register
may not persist when the device clock is disabled on some chips.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

Desvauxm-st commented:
LGTM for me

Desvauxm-st reviewed (approved):
OK for me

GeorgeCGV reviewed (approved):
Tested on H730, the `sys_rand32_get()` no longer fails.

niedzwiecki-dawid commented:
Thanks for all testing and sorry for the regression.
